app/queries/drug_stocks_query.rb

```diff
@@ -0,0 +1,117 @@
+class DrugStocksQuery
+  CACHE_VERSION = 1
+
+  def initialize(facilities:, for_end_of_month:)
+    @facilities = facilities
+    @for_end_of_month = for_end_of_month
+    # assuming that all facilities on the page have the same protocol
+    @protocol = @facilities.first.protocol
+    @state = @facilities.first.state
+  end
+
+  def protocol_drugs_by_category
+    @protocol_drugs_by_category ||= @protocol.protocol_drugs
+      .where(stock_tracked: true)
+      .sort_by(&:sort_key)
+      .group_by(&:drug_category)
+      .sort_by { |(drug_category, _)| drug_category }
+      .to_h
+  end
+
+  def cache_key
+    [
+      self.class.name,
+      @facilities.map(&:id).sort,
+      @for_end_of_month,
+      @protocol.id,
+      @state,
+      CACHE_VERSION
+    ].join("/")
+  end
+
+  def call
+    Rails.cache.fetch(cache_key,
+      expires_in: ENV.fetch("ANALYTICS_DASHBOARD_CACHE_TTL"),
+      force: RequestStore.store[:force_cache]) do
+      {all: totals,
+       facilities: report_for_facilities,
+       last_updated_at: Time.now}
+    end
+  end
+
+  def patient_counts
+    @patient_counts ||= Patient.where(assigned_facility_id: @facilities).group(:assigned_facility).count
+  end
+
+  def drug_stocks
+    @drug_stocks ||= DrugStock.latest_for_facilities(@facilities, @for_end_of_month).group_by(&:facility_id)
+  end
+
+  def report_for_facilities
+    @report_for_facilities ||= @facilities.each_with_object({}) { |facility, report|
+      report[facility.id] = drug_stock_for_facility(facility, drug_stocks[facility.id])
+    }
+  end
+
+  def drug_stock_for_facility(facility, facility_drug_stocks)
+    patient_count = patient_counts[facility] || 0
+    facility_report = {facility: facility, patient_count: patient_count}
+    return facility_report if facility_drug_stocks.nil?
+
+    drug_stocks = facility_drug_stocks.group_by { |drug_stock| drug_stock.protocol_drug.drug_category }
+    protocol_drugs_by_category.each do |(drug_category, _protocol_drugs)|
+      facility_report[drug_category] = drug_stock_for_category(drug_category, drug_stocks, patient_count)
+    end
+    facility_report
+  end
+
+  def drug_stock_for_category(drug_category, drug_stocks, patient_count)
+    patient_days = Reports::PatientDaysCalculation.new(
```

Comment:
I stared at this method for a little bit wondering why we return _nothing_ if there is no patient days information to provide. I then realized, this service puts together the full drug stock info - how many of each drug left + how many patient days that comes to.

Can we rename the service to something like `Reports::DrugStockCalculation` to make these responsibilities more evident?

Reply:
It actually returns the "calculation" of patient days, which includes all the factors that go into that equation and are then shown in the tooltip. The rest of the drug stock info on the report does not actually come from here.

However, we'll change around these lines with the consumption report since we'll reuse the class for that as well.